POLIO-2019 doses per vial

quang-le opened the pull request:
What problem is this PR solving? Explain here in one sentence.

Related JIRA tickets :  POLIO-2019, POLIO-1998

## Self proofreading checklist

- [x] Did I use eslint and ruff formatters?
- [x] Is my code clear enough and well documented?
- [x] Are my typescript files well typed?
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests?
- [ ] Documentation has been included (for new feature)

## Doc

[Will edit this section when doc is written](https://github.com/BLSQ/polio-doc/pull/4)

## Changes

- Backend:
    - Add an endpoint returning the possible presentations (doses_per_vial) based on a `Config` object. The slug of the config is hard coded/stored in a constant. For use in Supply chain, where we need to show all possible options
    - Add an endpoint returning the possible presentations and the number of usable doses and unusable doses for each presentation. For use in vaccine stock, where some actions are disabled if there's no stock available
    - Add methods to `VaccineStockCalculator` to compute the totals needed for the new API

- Front-end
    - SupplyChain:
        -  PreAlert: 
            - Replace `NumberInput` with `SingleSelect`
            - Set a `defaultDosesPerVial` only if there's only 1 possible presentation
            - Make doses per vial uneditable if there's only one presentation
        - Arrival Report: same as PreAlert, plus:
            - auto-fill `doses_per_vial` when user picks a PO number from the auto-generated list (from pre-alerts)
     - Vaccine Stock:
         - FormA, Earmarked:
             - disable creation if there's no stock of usable doses
              - Replace `NumberInput` with `SingleSelect`
              - Only show presentations for which stock of usable doses > 0 in the dropdown options
              - Edition: add the saved value for presentation/doses per vial to the dropdown to avoid having an errored UI
          - Destruction reports:
              - Same as FormA and Earmarked, but based on stock of unusable doses
          - Incident reports:
              - Never disable Creation, to allow for encoding of Physical Inventory
              - Disable movements from usable stock to unusable if usable stock <= 0
              - Physical Inventory: disable some options based on usable/unusable stock being available 
              
## How to test

With a DB setup for polio, go to the django admin and add a `Config` with the following:
slug=vaccine_doses_per_vial
content=
```json
{
  "bOPV": [
    10,
    20
  ],
  "mOPV2": [
    20,
    50
  ],
  "nOPV2": [
    50
  ]
}
```

Then go to supply chain:
- Try adding a PreAlert: change the values --> the doses and vial values should update. Saving should work OK
- Try adding a vaccine arrival report: select a PO number from the dropdown --> The doses per vial should auto fill

Go to vaccine stock:
- Try adding/editing each type of report for every vaccine. nOPV2 should have a different behaviour: the dropdowns should be (pre)filled with 50 and not be editable

## Print screen / video

https://github.com/user-attachments/assets/80d23082-999c-4d8f-9164-aa41717f8711

https://github.com/user-attachments/assets/7ab8affd-3905-459d-a9ff-429efb23a1e8

quang-le committed:
POLIO-2019: add API to get doses per vial

    - based on hard-coded Config slug
    - read only
    - require polio or vaccine perm

quang-le committed:
POLIO-2019: add doses per vial endpoints

- add doses_options endpoitn that returns presentations for which there is stock (to use in vaccine stock)
- add doses_per_vial endpoint with all accepted prsentations per vaccine (to use in supply chain)

quang-le committed:
POLIO-2019: turn doses per vial to Select in supply chain

- PreAlerts: fetch presentations per vaccine, use them to populate dropdown
- Arrival reports: same, but also auto-fill doses per vial when selecting a PO from a pre alert
- both: disable and auto-fill field if only one presentation is available

quang-le committed:
POLIO-2019: fix tests

quang-le committed:
POLIO-2019: update bluesaquare-components

quang-le committed:
POLIO-2019: add unusable stock to API

quang-le committed:
POLIO-2019: add constraints on doses in vaccine stock front end

Bewi commented on hat/assets/js/apps/Iaso/components/forms/InputComponent.tsx:
```diff
@@ -284,6 +284,8 @@ const InputComponent: React.FC<InputComponentProps> = ({
                         dataTestId={dataTestId}
                         placeholder={placeholder}
+                        onBlur={onBlur}
+                        // onFocus={onFocus}
```
Clean up comment

Bewi commented on plugins/polio/api/urls.py:
```diff
@@ -110,6 +111,7 @@
 router.register(r"polio/tasks/refreshim/hh_ohh", RefreshIMAllDataViewset, basename="refreshimhhohh")
 router.register(r"polio/vaccine/request_forms", VaccineRequestFormViewSet, basename="vaccine_request_forms")
 router.register(r"polio/vaccine/vaccine_stock", VaccineStockManagementViewSet, basename="vaccine_stocks")
+router.register(r"polio/vaccine/doses_per_vial", DosesPerVaccineViewset, basename="doses_per_vial")
```
Maybe rename the views to match the route ?

Bewi commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditDestruction.tsx:
```diff
@@ -45,12 +50,15 @@ export const CreateEditDestruction: FunctionComponent<Props> = ({
+
+    const defaultDosesPerVial =
+        //@ts-ignore
+        (dosesOptions ?? []).length === 1 ? dosesOptions[0].value : undefined;
```
It feels a bit hacky, wouldn't it be better to have a default on dosesOptions ?

Bewi replied:
Or to not allow null and let the parent define [] if dosesOptions is undefined

quang-le replied:
We only want to preset the value in the UI if there's 1 option, and we want to use the value sent by the API, so this pretty much spells out the business logic. The only reason there's a null check on `dosesOptions` is to prevent errors in the UI while data is being fetched

quang-le committed:
POLIO-2019: fix value bug in VAR

Bewi commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/Modals/CreateEditEarmarked.tsx:
```diff
@@ -76,7 +85,24 @@ export const CreateEditEarmarked: FunctionComponent<Props> = ({
     );
     const earmarkTypeOptions = useEarmarkOptions();
-
+    const availableDosesPresentations: DropdownOptions<number>[] =
```
Feels like this is 99% the same as availableDosesPresentations of CreateEditDestruction, do you think it would be possible to have it in one place with a param for the key to use (doses_available) and earmark/destruction as param ?

quang-le replied:
Good catch, I'll try it, see if I don't get weird state synchro bugs

Bewi commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/VaccineStockVariation.tsx:
```diff
@@ -72,6 +73,11 @@ export const VaccineStockVariation: FunctionComponent = () => {
     });
 
+    const { data: dosesOptions } = useGetDosesOptions(parseInt(params.id, 10));
+    const hasUsableStock =
```
Would a useMemo make sense here ?

quang-le replied:
No, because it's a primitive (boolean) so there's no reference trap

Bewi commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/hooks/api.ts:
```diff
@@ -253,9 +253,18 @@ export const useGetEarmarkedList = (
+export const useGetDosesOptions = (
+    stockId: number,
+): UseQueryResult<
+    {
```
This is used in multiple places, a type would help to prevent misalignment in the future

Bewi reviewed (commented):
This is a first review for the FE mainly, I still need to run it locally

tdethier commented on plugins/polio/models/base.py:
```diff
@@ -2792,3 +2795,61 @@ def get_list_of_earmarked(self, expanded=False):
+                    total_unusable_vials -= vial["vials_out"]
+                if vial["doses_out"]:
+                    total_unusable_doses -= vial["doses_out"]
+        return total_unusable_vials, total_unusable_doses
```
We can probably refactor both into a single method and pass a parameter to determiner which method should be call to fetch doses and vials data

tdethier commented on plugins/polio/models/base.py:
```diff
@@ -2792,3 +2795,61 @@ def get_list_of_earmarked(self, expanded=False):
+
+        options = presentation_config.content[self.vaccine_stock.vaccine]
+        if not options:
+            return None
+        results = {}
+        for option in options:
+            results[str(option)] = self._get_usable_stock_for_presentation(option)[1]
+        return results
+
+    def get_unusable_stock_by_vaccine_presentation(self):
+        presentation_config = Config.objects.filter(slug=DOSES_PER_VIAL_CONFIG_SLUG).first()
+        if not presentation_config:
+            return None
+
+        options = presentation_config.content[self.vaccine_stock.vaccine]
+        if not options:
+            return None
+        results = {}
+        for option in options:
+            results[str(option)] = self._get_unusable_stock_for_presentation(option)[1]
+        return results
```
If `self._get_(un)usable_stock_for_presentation` is refactored, you can easily extract the code of both into a single method and pass a parameter to select usuable or unsuable because here again the content is the same

tdethier commented on plugins/polio/api/vaccines/doses_per_vaccine.py:
```diff
@@ -0,0 +1,54 @@
+        ),
+    ]
+    serializer_class = ConfigSerializer
+    CONFIG_SLUG = DOSES_PER_VIAL_CONFIG_SLUG
```
Nitpick: this works fine, but I would have directly used the constant instead of defining a new constant with the contstant

quang-le replied:
Oh yeah, that's a leftover from dev

tdethier commented on plugins/polio/tests/api/vaccine_stock_management/test_doses_per_vaccine_endpoint.py:
```diff
@@ -0,0 +1,49 @@
+    def test_anonymous_user_cannot_list(self):
+        self.client.force_authenticate(self.anon)
+        response = self.client.get(BASE_URL)
+        self.assertEqual(response.status_code, 403)
```
It's weird that it's not a 401 - I guess it happens if you don't `force_authenticate` at all

tdethier commented on plugins/polio/tests/api/vaccine_stock_management/test_doses_per_vaccine_endpoint.py:
```diff
@@ -0,0 +1,49 @@
+    def test_user_without_perms_cannot_list(self):
+        self.client.force_authenticate(self.user_no_perms)
+        response = self.client.get(BASE_URL)
+        self.assertEqual(response.status_code, 403)
```
Nitpick: I prefer using constants from `rest_framework.status` instead of `int` values

tdethier commented on plugins/polio/tests/test_vaccine_stock_management.py:
```diff
@@ -1736,225 +1748,146 @@ def test_doses_options_endpoint_invalid_stock_id(self):
+        # We expect some unusable doses for 20-dose vials based on our test data
+        item_20 = next((item for item in results if item["value"] == 20), None)
+        self.assertIsNotNone(item_20)
+        self.assertGreaterEqual(item_20["unusable_doses"], 0)
```
Why `assertGreaterEqual` and not `assertEqual` with the right value?

quang-le committed:
POLIO-2019: improve TS typing

quang-le committed:
POLIO-2019: rename Viewset

quang-le committed:
POLIO-2019: fix bugs in VAR form

- prevent data overwrite from PreAlert (pre-fill bug)
- correctly use PreAlert value on first pre-fill

tdethier commented on plugins/polio/api/vaccines/doses_per_vaccine.py:
```diff
@@ -0,0 +1,54 @@
+    CONFIG_SLUG = DOSES_PER_VIAL_CONFIG_SLUG
+
+    def get_queryset(self):
+        return Config.objects.filter(slug=self.CONFIG_SLUG)
```
I'm still not a big fan of `Config`, but I guess this works fine 😅

quang-le replied:
Yep, arbitration between time available and need for configurability

quang-le committed:
POLIO-2019: improve handling of single option in stock

- if only one option: the doses_per_vial is pre-filled and disabled
- the default doses per vial is passed from the parent

tdethier reviewed (approved):
The backend looks good, but I left a few comments.

I tested the feature, it works fine 👍 

I have a couple of suggestions regarding the user experience:
1. when I create a new form A for a nOPV2 vaccine, the `50` value is hardcoded, but it generates a validation error. It stills works, but it's weird to have the error straight away.
<img width="660" height="610" alt="Image" src="https://github.com/user-attachments/assets/db985a5b-b7e4-453f-887a-92b1bea825cf" />
 (by the way, the calls made to the backend to load dropdown values are crazy 😵‍💫 

```
http://localhost:8081/api/polio/campaigns/?campaign_category=all&show_test=true&on_hold=true&is_embedded=false&enabled=true
|------|-----------|----------|----------|----------|------------|
| Type | Database  |   Reads  |  Writes  |  Totals  | Duplicates |
|------|-----------|----------|----------|----------|------------|
| RESP |  default  |   3314   |    0     |   3314   |    2864    |
|------|-----------|----------|----------|----------|------------|
Total queries: 3314 in 45.1416s 
```

2. when I prepare a new pre alert, I would expect the order of fields to be the opposite:
<img width="1783" height="424" alt="Image" src="https://github.com/user-attachments/assets/9f8300f7-099d-4ed1-a1ed-2c23b51172e4" />
Since the doses per vial change the other fields, it makes sense to fill it out first and then do the others in that order

The vaccine arrival reports form has a similar issue.

quang-le committed:
POLIO-2019: update queryKeys invalidations

quang-le committed:
POLIO-2019: codereview (backend)

quang-le committed:
POLIO-2019: code review (front) + bug fixes

- fix dropdown options for incident report
- add useAvailablePresentations hook

Bewi commented on plugins/polio/js/src/domains/VaccineModule/StockManagement/StockVariation/VaccineStockVariation.tsx:
```diff
@@ -112,7 +112,7 @@ export const VaccineStockVariation: FunctionComponent = () => {
         summary?.vaccine_type,
     );
-
+    console.log('HAS USABLE', hasUsableStock, dosesOptions, tab);
```
Clean up :D

tdethier reviewed (approved):
Backend looks good, thanks for the changes!

quang-le committed:
chore: clean up console.log
